icu4c/source/common/cjmlbe.h

```diff
@@ -0,0 +1,126 @@
+// © 2022 and later: Unicode, Inc. and others.
+// License & terms of use: http://www.unicode.org/copyright.html
+
+#ifndef CJMLBREAKENGINE_H
+#define CJMLBREAKENGINE_H
+
+#include "unicode/utypes.h"
+#include "unicode/uniset.h"
+#include "unicode/utext.h"
+
+#include "brkeng.h"
+#include "dictbe.h"
+#include "hash.h"
+#include "uvectr32.h"
+
+
+U_NAMESPACE_BEGIN
+
+#if !UCONFIG_NO_NORMALIZATION
+
+
+/**
+ * A machine learning break engine for the phrase breaking in Chinese and Japanese.
+ */
+class CjMLBreakEngine : public UMemory {
+
+public:
+    /**
+     * Constructor.
+     *
+     * @param fDigitOrOpenPunctuationOrAlphabetSet A UnicodeSet with Digit, open punctuation and alphabet.
+     * @param fClosePunctuationSet A UnicodeSet with close punctuation.
+     * @param status Information on any errors encountered.
+     */
+    CjMLBreakEngine(UnicodeSet &fDigitOrOpenPunctuationOrAlphabetSet, UnicodeSet &fClosePunctuationSet, UErrorCode &status);
+
+    /**
+     * <p>Virtual destructor.</p>
+     */
+    virtual ~CjMLBreakEngine();
+
+public:
+    /**
+     * Divide up a range of characters handled by this break engine.
+     *
+     * @param inText A UText representing the text
+     * @param rangeStart The start of the range of dictionary characters
+     * @param rangeEnd The end of the range of dictionary characters
+     * @param foundBreaks Output of C array of int32_t break positions, or 0
+     * @param inString The normalized string of text ranging from rangeStart to rangeEnd
+     * @param inputMap The vector storing the native index of inText
+     * @param status Information on any errors encountered.
+     * @return The number of breaks found
+     */
+    int32_t divideUpDictionaryRange(UText *inText,
+                                            int32_t rangeStart,
+                                            int32_t rangeEnd,
+                                            UVector32 &foundBreaks,
+                                            UnicodeString &inString,
+                                            LocalPointer<UVector32> &inputMap,
+                                            UErrorCode& status) const;
+
+private:
+
+    /**
+     * Load the machine learning's model file.
+     *
+     * @param error Information on any errors encountered.
+     */
+    void loadMLparams(UErrorCode& error);
+    /**
+     * Get the unicode block index of the unicode character.
+     *
+     * @param ch A unicode character.
+     * @param error Information on any errors encountered.
+     * @return The unicode block index
+     */
+     UnicodeString getUnicodeBlockIndex(UChar32 ch, UErrorCode& status) const;
+
+    /**
+     * Initiate the element list from the input string.
+     *
+     * @param inString A input string to be segmented
+     * @param elementList A vector to store the first six characters and their unicode block indexes.
```

Comment:
not clear after this call, is the CjMLBreakEngine owning the memory of elementList and response to free it or not.  @markusicu  does it matter this is a const pointer or not?

Reply:
The vector should be const if the caller keeps ownership and the callee is not supposed to modify it. If the callee is to "adopt" the vector, then it should be non-const, and the ownership transfer should be documented. Also, adopting an object needs to happen consistently, even when an error occurs. If in doubt, have the function immediately capture the adopted object in a LocalPointer.

Reply:
In this function, elementList is initiated by adding six objects so that it could not be const.